File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    d: dict[str, str] = {
        "2": ["a","b", "c"],
        "3": ["d","e", "f"],
        "4": ["g","h", "i"],
        "5": ["j","k", "l"],
        "6": ["m","n", "o"],
        "7": ["p","q", "r", "s"],
        "8": ["t","u", "v"],
        "9": ["w","x", "y", "z"],
    }

    def letterCombinations(self, digits: str) -> List[str]:
        return self.build_letter_combos(digits)

    def build_letter_combos(self, digits: str) -> List[str]:
        letter_combos: List[str] = []
        
        if len(digits) == 0: return letter_combos

        curr_letter_combos = self.d[digits[0]]
        # print(curr_letter_combos)
        new_digits = digits[1:]
        # print(new_digits)

        next_letter_combos = self.build_letter_combos(new_digits)
        # print(next_letter_combos)
        if len(next_letter_combos) == 0: return curr_letter_combos

        for curr_letter_combo in curr_letter_combos:
            for next_letter_combo in next_letter_combos:
                letter_combos.append(curr_letter_combo + next_letter_combo)
        
        return letter_combos
        

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    static Map<String, String> PHONE_DIAL_MAP = new HashMap<> ();
    ArrayList<String> letters = new ArrayList<>();
    ArrayList<String> letterCombos = new ArrayList<>();

    static {
        PHONE_DIAL_MAP.put("2", "abc");
        PHONE_DIAL_MAP.put("3", "def");
        PHONE_DIAL_MAP.put("4", "ghi");
        PHONE_DIAL_MAP.put("5", "jkl");
        PHONE_DIAL_MAP.put("6", "mno");
        PHONE_DIAL_MAP.put("7", "pqrs");
        PHONE_DIAL_MAP.put("8", "tuv");
        PHONE_DIAL_MAP.put("9", "wxyz");
    }

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return letterCombos;
        }

        for (int i = 0; i < digits.length(); i ++) {
            letters.add(PHONE_DIAL_MAP.get(String.valueOf(digits.charAt(i))));
        }
        
        backtrack("", 0);
        return letterCombos;
    }

    public void backtrack(String currString, int i) {
        if (currString.length() == letters.size()) {
            letterCombos.add(currString);
            return;
        }

        String currLetters = letters.get(i);

        for (int j = 0; j < currLetters.length(); j++) {
            String newCurrString = currString + currLetters.charAt(j);
            backtrack(newCurrString, i + 1);
        }

    }


    
}